### PR TITLE
[Merged by Bors] - feat(Data/Nat/Factorial): add descFactorial_le and descFactorial_mul_descFactorial

### PR DESCRIPTION
**Pull request by nmouha**
Add two basic theorems involving falling factorials. Note that `descFactorial_mul_descFactorial` collapses to `Nat.factorial_mul_descFactorial` when `m = n`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Nat/Factorial/Basic.lean
Comment (ScottCarnahan) [inline Mathlib/Data/Nat/Factorial/Basic.lean:383]:
```suggestion
theorem descFactorial_mul_descFactorial {k m n : ℕ} (hkm : k ≤ m) : 
    (n - k).descFactorial (m - k) * n.descFactorial k = n.descFactorial m := by
  by_cases hmn : m ≤ n
  · apply Nat.mul_left_cancel (n - m).factorial_pos
    rw [factorial_mul_descFactorial hmn, show n - m = (n - k) - (m - k) by omega, ← Nat.mul_assoc,
      factorial_mul_descFactorial (show m - k ≤ n - k by omega),
      factorial_mul_descFactorial (le_trans hkm hmn)]
  · rw [descFactorial_eq_zero_iff_lt.mpr (show n < m by omega)]
    by_cases hkn : k ≤ n
    · rw [descFactorial_eq_zero_iff_lt.mpr (show n - k < m - k by omega), Nat.zero_mul]
    · rw [descFactorial_eq_zero_iff_lt.mpr (show n < k by omega), Nat.mul_zero]
```

You have a nice proof by induction, but in some sense, the proof of `Nat.factorial_mul_descFactorial` has already done the induction for you.

File: Mathlib/Data/Nat/Factorial/Basic.lean
Comment (ScottCarnahan) [inline Mathlib/Data/Nat/Factorial/Basic.lean:391]:
```suggestion
theorem descFactorial_le (n : ℕ) {k m : ℕ) (h : k ≤ m) :
```

You don't need `k` and `m` to be explicit if they are related by an explicit inequality.

**Review by ScottCarnahan — COMMENTED** [review]
Thanks for contributing!

**Comment by nmouha** [comment]
Thank you, @ScottCarnahan, for the extremely quick reply! I was expecting that the proof could be improved but I didn't know how.

Your proof clearly looks much cleaner, and I really appreciate that you're taking time to explain all of this to me.

I've accepted all your suggestions (with extremely minor changes just to avoid lint-style and build errors). Also, I've replaced `exact Nat.le_of_ble_eq_true rfl` by `rfl`. (In hindsight, Lean's `exact?` made a correct but unnecessarily complicated suggestion here.)

If anything further is required on my side before this PR can be merged, just let me know!

File: Mathlib/Data/Nat/Factorial/Basic.lean
Comment (ScottCarnahan) [inline Mathlib/Data/Nat/Factorial/Basic.lean:383]:
```suggestion
    exact Nat.mul_le_mul (Nat.sub_le_sub_right h n) ih
```

**Review by ScottCarnahan — APPROVED** [review]
Looks good to me.

**Review by Ruben-VandeVelde — APPROVED** [review]
maintainer merge

**Comment by mattrobball** [comment]
bors merge

**Comment by bryangingechen** [comment]
One of the runners ran out of space. Re-running
bors r+

**Pull request closed without merging**
